File: solutions/lua/luhn/1/luhn.lua
return {
  valid = function(s)
    if select(2, s:gsub('[%s%d]', '')) ~= #s then return false end
    local sum = 0
    local double = false
    for i in s:reverse():gmatch('%d') do
      n = tonumber(i)
       if double then
        n = n * 2
        if n > 9 then
          n = n - 9
        end
      end
      sum = sum + n
      double = not double
    end
    return sum % 10 == 0 and (
      sum > 0 or select(2, s:gsub('[%d]', '')) > 1
    )
  end
}

File: solutions/go/luhn/1/luhn.go
package luhn

func Valid(id string) bool {
	var sum int
	double := false
	for i := len(id) - 1; i >= 0; i-- {
		r := id[i]

		if r == ' ' {
			continue
		}

		if r < '0' || r > '9' {
			return false
		}

		n := int(r - '0')
		if double {
			n *= 2
			if n > 9 {
				n -= 9
			}
		}
		double = !double

		sum += n
	}

	return sum%10 == 0 && sum != 0
}
